<a href="https://colab.research.google.com/github/Ruthmwaura/module_2_week_4_ip/blob/main/Naive_Bayes_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

a) Specifying the Data Analytic Question

predicting whether the email is either spam (1) or not (0).

b) Defining the Metric for Success

Success will be pegged on the ability to create a model that has high accuracy on classifying the type of email.

*c) Understanding the context *

it is annoying to have an inbox full of spam mails. it is therefore neccessary to have a way to tell if an email is spam or not and the spam mails can be saved on a different folder.

d) Recording the Experimental Design

To successfully conduct this project the we will use CRISP DM, following steps will be taken in order:

-Business understanding

-Data understanding

-Data preparation

-explanatory data analysis

-modelling

-evaluation

-challenging solution

e)Data appropriateness

The data have variables that will adequately give answers to the analytical question

In [1]:
#imports
import  pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB

In [2]:
columns=['word_freq_make','word_freq_address','word_freq_all','word_freq_3d','word_freq_our','word_freq_over','word_freq_remove','word_freq_internet',
'word_freq_order','word_freq_mail','word_freq_receive','word_freq_will','word_freq_people','word_freq_report','word_freq_addresses','word_freq_free', 
'word_freq_business','word_freq_email','word_freq_you','word_freq_credit','word_freq_your','word_freq_font','word_freq_000','word_freq_money',        
'word_freq_hp','word_freq_hpl','word_freq_george','word_freq_650','word_freq_lab','word_freq_labs','word_freq_telnet','word_freq_857','word_freq_data',         
'word_freq_415','word_freq_85','word_freq_technology','word_freq_1999','word_freq_parts','word_freq_pm','word_freq_direct','word_freq_cs','word_freq_meeting',      
'word_freq_original','word_freq_project','word_freq_re','word_freq_edu','word_freq_table','word_freq_conference','char_freq1','char_freq2','char_freq3',            
'char_freq4','char_freq5','char_freq6','capital_run_length_average','capital_run_length_longest', 'capital_run_length_total','spam'] 

df=pd.read_csv('/content/spambase.data',names=columns)


#Data Understanding

In [3]:
#previewing the 1st 5 columns
df.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq1  char_freq2  char_freq3  \
0             0.00            0.00  ...        0.00       0.000         0.0   
1             0.00            0.94  ...        0.00       0.132         0.0   
2             0.64            0.25  ...        0.01       0.143         0.0   
3             0.31            0.63  ...        0.00       0.137         0.0   
4             0.31            0.63  ...        0.00       0.135         0.0   

   char_freq4  char_freq5  char_freq6  capital_run_length_average  \
0       0.778       0.000       0.000                       3.756   
1       0.372       0.180       0.048                       5.114   
2       0.276       0.184       0.010                       9.821   
3       0.137       0.000       0.000                       3.537   
4       0.135       0.000       0.000                       3.537   

   capital_run_length_longest  capital_run_length_total  spam  
0                          61                       278     1  
1                         101                      1028     1  
2                         485                      2259     1  
3                          40                       191     1  
4                          40                       191     1  

[5 rows x 58 columns]

In [4]:
#previewing the last 5 columns
df.tail()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4596            0.31                0.0           0.62           0.0   
4597            0.00                0.0           0.00           0.0   
4598            0.30                0.0           0.30           0.0   
4599            0.96                0.0           0.00           0.0   
4600            0.00                0.0           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4596           0.00            0.31               0.0                 0.0   
4597           0.00            0.00               0.0                 0.0   
4598           0.00            0.00               0.0                 0.0   
4599           0.32            0.00               0.0                 0.0   
4600           0.00            0.00               0.0                 0.0   

      word_freq_order  word_freq_mail  ...  char_freq1  char_freq2  \
4596              0.0             0.0  ...       0.000       0.232   
4597              0.0             0.0  ...       0.000       0.000   
4598              0.0             0.0  ...       0.102       0.718   
4599              0.0             0.0  ...       0.000       0.057   
4600              0.0             0.0  ...       0.000       0.000   

      char_freq3  char_freq4  char_freq5  char_freq6  \
4596         0.0       0.000         0.0         0.0   
4597         0.0       0.353         0.0         0.0   
4598         0.0       0.000         0.0         0.0   
4599         0.0       0.000         0.0         0.0   
4600         0.0       0.125         0.0         0.0   

      capital_run_length_average  capital_run_length_longest  \
4596                       1.142                           3   
4597                       1.555                           4   
4598                       1.404                           6   
4599                       1.147                           5   
4600                       1.250                           5   

      capital_run_length_total  spam  
4596                        88     0  
4597                        14     0  
4598                       118     0  
4599                        78     0  
4600                        40     0  

[5 rows x 58 columns]

In [5]:
#checking number of columns and rows
df.shape

(4601, 58)

the data has 4,601 rows and 58 columns

In [6]:
#checking the datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

#Data preparation

##validity


all columns will be used

##accuracy


we will assume that the data provided is accurate

##consistency

we will look for duplicates

In [7]:
#checking for duplicates
df.duplicated().sum()

391

In [8]:
#dropping the duplicates
df.drop_duplicates(subset=None, keep='first', inplace=True, ignore_index=False)

##completeness

we will check for null values

In [9]:
df.isnull().any()

word_freq_make                False
word_freq_address             False
word_freq_all                 False
word_freq_3d                  False
word_freq_our                 False
word_freq_over                False
word_freq_remove              False
word_freq_internet            False
word_freq_order               False
word_freq_mail                False
word_freq_receive             False
word_freq_will                False
word_freq_people              False
word_freq_report              False
word_freq_addresses           False
word_freq_free                False
word_freq_business            False
word_freq_email               False
word_freq_you                 False
word_freq_credit              False
word_freq_your                False
word_freq_font                False
word_freq_000                 False
word_freq_money               False
word_freq_hp                  False
word_freq_hpl                 False
word_freq_george              False
word_freq_650               

we do not have null values

##uniformity

we check for outliers and anomalies.

In [10]:
#checking for outliers
numerical_data= df.select_dtypes(exclude='category')

Q1 = numerical_data.quantile(0.25)
Q3 = numerical_data.quantile(0.75)
IQR = Q3 - Q1

(numerical_data < (Q1 - 1.5 * IQR)) |(numerical_data > (Q3 + 1.5 * IQR))


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0              False               True          False         False   
1               True               True          False         False   
2               True              False          False         False   
3              False              False          False         False   
4              False              False          False         False   
...              ...                ...            ...           ...   
4596            True              False          False         False   
4597           False              False          False         False   
4598            True              False          False         False   
4599            True              False          False         False   
4600           False              False          False         False   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0             False           False             False               False   
1             False            True              True                True   
2              True            True              True                True   
3             False           False              True                True   
4             False           False              True                True   
...             ...             ...               ...                 ...   
4596          False            True             False               False   
4597          False           False             False               False   
4598          False           False             False               False   
4599          False           False             False               False   
4600          False           False             False               False   

      word_freq_order  word_freq_mail  ...  char_freq1  char_freq2  \
0               False           False  ...       False       False   
1               False            True  ...       False       False   
2                True           False  ...        True       False   
3                True            True  ...       False       False   
4                True            True  ...       False       False   
...               ...             ...  ...         ...         ...   
4596            False           False  ...       False       False   
4597            False           False  ...       False       False   
4598            False           False  ...        True        True   
4599            False           False  ...       False       False   
4600            False           False  ...       False       False   

      char_freq3  char_freq4  char_freq5  char_freq6  \
0          False       False       False       False   
1          False       False        True        True   
2          False       False        True        True   
3          False       False       False       False   
4          False       False       False       False   
...          ...         ...         ...         ...   
4596       False       False       False       False   
4597       False       False       False       False   
4598       False       False       False       False   
4599       False       False       False       False   
4600       False       False       False       False   

      capital_run_length_average  capital_run_length_longest  \
0                          False                       False   
1                          False                        True   
2                           True                        True   
3                          False                       False   
4                          False                       False   
...                          ...                         ...   
4596                       False                       False   
4597                       False                       False   
4598                       False                       False   
4599                       False                       False  

we will not drop the outliers.

#EDA Analysis

In [11]:
#checking the correlaton
df.corr()

word_freq_make  word_freq_address  word_freq_all  \
word_freq_make                    1.000000           0.034012       0.063207   
word_freq_address                 0.034012           1.000000       0.027789   
word_freq_all                     0.063207           0.027789       1.000000   
word_freq_3d                      0.005388          -0.008947      -0.019901   
word_freq_our                     0.021740           0.036975       0.065676   
word_freq_over                    0.055067           0.005526       0.066547   
word_freq_remove                  0.011061           0.077374       0.028581   
word_freq_internet               -0.004370           0.013706       0.007221   
word_freq_order                   0.103842           0.053426       0.077561   
word_freq_mail                    0.040829           0.184304       0.025138   
word_freq_receive                 0.063031           0.060724       0.055647   
word_freq_will                    0.075911           0.026247       0.065733   
word_freq_people                  0.061985           0.015538       0.036431   
word_freq_report                  0.032263           0.012414       0.004900   
word_freq_addresses               0.022454           0.061826       0.104574   
word_freq_free                    0.065443           0.046630       0.061937   
word_freq_business                0.081471           0.022018       0.029834   
word_freq_email                   0.051571           0.178853       0.110794   
word_freq_you                     0.120563           0.055876       0.126502   
word_freq_credit                  0.020924          -0.006735       0.030874   
word_freq_your                    0.121458           0.117797       0.147895   
word_freq_font                   -0.025574           0.001214      -0.039464   
word_freq_000                     0.126478           0.007064       0.106694   
word_freq_money                   0.205689           0.061282       0.031150   
word_freq_hp                     -0.075054          -0.053573      -0.092684   
word_freq_hpl                    -0.063467          -0.044627      -0.066464   
word_freq_george                 -0.048086          -0.011228      -0.081297   
word_freq_650                    -0.050054          -0.033000      -0.051471   
word_freq_lab                    -0.041541          -0.026048      -0.057847   
word_freq_labs                   -0.054241          -0.029419      -0.030519   
word_freq_telnet                 -0.038899          -0.015937      -0.036610   
word_freq_857                    -0.031634           0.028852      -0.063579   
word_freq_data                   -0.042715          -0.032575      -0.058002   
word_freq_415                    -0.026322           0.026039      -0.063417   
word_freq_85                     -0.046083          -0.025524      -0.048968   
word_freq_technology             -0.057296          -0.028262      -0.047593   
word_freq_1999                   -0.058771           0.003749      -0.074329   
word_freq_parts                  -0.008626          -0.012708       0.031756   
word_freq_pm                     -0.011199          -0.016208      -0.017418   
word_freq_direct                 -0.035467           0.002326      -0.046013   
word_freq_cs                     -0.005280          -0.019350      -0.030822   
word_freq_meeting                -0.026546          -0.035908      -0.007620   
word_freq_original               -0.025603           0.037681      -0.052736   
word_freq_project                -0.022980          -0.029055      -0.056883   
word_freq_re                     -0.037848           0.020029      -0.056600   
word_freq_edu                    -0.034455          -0.026004      -0.060558   
word_freq_table                  -0.000950          -0.012337       0.027384   
word_freq_conference             -0.018904          -0.020947      -0.029465   
char_freq1                       -0.027248           0.014735      -0.036778   
char_freq2                  

using the threshold of 0.5, there are no variables that are closely related.

In [12]:
#checking multi colinearity between the independent variables
#using correlations to compute the VIF score for each variable
independent=df.drop(['spam'], axis=1)
correlations=independent.corr()
pd.DataFrame(np.linalg.inv(correlations.values), index = correlations.index, columns=correlations.columns)

word_freq_make  word_freq_address  word_freq_all  \
word_freq_make                    1.101307          -0.004149      -0.015513   
word_freq_address                -0.004149           1.107902       0.003449   
word_freq_all                    -0.015513           0.003449       1.094293   
word_freq_3d                     -0.000943           0.015624       0.023958   
word_freq_our                     0.009130          -0.011665      -0.019600   
word_freq_over                   -0.012852           0.015299      -0.020634   
word_freq_remove                  0.035940          -0.039440       0.024997   
word_freq_internet                0.049783           0.014547       0.028332   
word_freq_order                  -0.061157          -0.006352      -0.023654   
word_freq_mail                   -0.005142          -0.189365       0.023129   
word_freq_receive                -0.002407           0.001803       0.001643   
word_freq_will                   -0.069767          -0.026909      -0.045850   
word_freq_people                 -0.019472           0.011289      -0.004460   
word_freq_report                 -0.008093          -0.011133       0.009541   
word_freq_addresses               0.054184           0.007194      -0.059010   
word_freq_free                   -0.031635          -0.013729      -0.030515   
word_freq_business               -0.052882           0.017375       0.027109   
word_freq_email                  -0.024420          -0.195293      -0.063662   
word_freq_you                    -0.047628           0.021102      -0.071539   
word_freq_credit                  0.017752           0.028145      -0.003110   
word_freq_your                   -0.035872          -0.067829      -0.088681   
word_freq_font                    0.021354           0.012870       0.039762   
word_freq_000                    -0.106550           0.043038      -0.050933   
word_freq_money                  -0.181826          -0.026421       0.027965   
word_freq_hp                      0.017743           0.026019       0.050296   
word_freq_hpl                     0.005897           0.017162       0.003073   
word_freq_george                  0.017008          -0.001674       0.052397   
word_freq_650                     0.000451           0.027832       0.007980   
word_freq_lab                     0.025073           0.024292       0.050844   
word_freq_labs                    0.018146           0.048690      -0.056212   
word_freq_telnet                 -0.004171           0.030031      -0.028826   
word_freq_857                     0.464784          -0.521923      -0.080209   
word_freq_data                    0.018447           0.012288       0.039043   
word_freq_415                    -0.540509           0.283663       0.113481   
word_freq_85                      0.006108           0.005823       0.017970   
word_freq_technology              0.026920           0.050034      -0.019227   
word_freq_1999                    0.028469          -0.025261       0.023059   
word_freq_parts                   0.000671           0.011170      -0.033002   
word_freq_pm                     -0.016170           0.016371      -0.015303   
word_freq_direct                  0.040049           0.074208       0.023666   
word_freq_cs                     -0.024390           0.005907      -0.000156   
word_freq_meeting                -0.002442           0.007657      -0.029411   
word_freq_original               -0.013158          -0.042277       0.017639   
word_freq_project                 0.006627           0.015813       0.046810   
word_freq_re                      0.012635          -0.041054       0.041781   
word_freq_edu                     0.015534           0.012958       0.044995   
word_freq_table                   0.000030           0.015866      -0.029645   
word_freq_conference              0.005559           0.013124       0.017771   
char_freq1                        0.009466          -0.028008       0.009822   
char_freq2                  

there is no multicollinearity for there is no variable with a score above 5

#Naive Bayes Classifier

##80-20

###Gaussian Naive Bayes Classifier

In [13]:
#dividing data into dependent and independent variable
X=df.drop(['spam'], axis=1).values
y=df['spam'].values

#splitting data into train and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=6) 

#defining the model
clf = GaussianNB()  

#training the model
model1 = clf.fit(X_train, y_train) 

#making predictions
y_pred = model1.predict(X_test)

#creating a comparison dataframe

comparison_frame = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})

comparison_frame.head(5)

Actual  Predicted
0       1          1
1       1          1
2       0          0
3       0          0
4       0          0

In [14]:
#evaluating the model
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))


[[368 147]
 [ 12 315]]
              precision    recall  f1-score   support

           0       0.97      0.71      0.82       515
           1       0.68      0.96      0.80       327

    accuracy                           0.81       842
   macro avg       0.83      0.84      0.81       842
weighted avg       0.86      0.81      0.81       842



the model was ble to classify 368 and 315 messages correctly.

it classified 147 and 12 messages wrongly.

the model accuracy is 81%,this is not bad.

###Multinomial Naive Bayes Classifier

In [17]:
#defining the model and training 
model2 = MultinomialNB().fit(X_train, y_train)

#making predictions
y_pred1 = model2.predict(X_test)

#creating a comparison dataframe

comparison_frame = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred1})

comparison_frame.head(5)

Actual  Predicted
0       1          0
1       1          0
2       0          0
3       0          0
4       0          0

In [18]:
#evaluating the model

print(confusion_matrix(y_test, y_pred1))
print(classification_report(y_test, y_pred1))


[[447  68]
 [111 216]]
              precision    recall  f1-score   support

           0       0.80      0.87      0.83       515
           1       0.76      0.66      0.71       327

    accuracy                           0.79       842
   macro avg       0.78      0.76      0.77       842
weighted avg       0.79      0.79      0.78       842



the accuracy dropped 

also this model classified lesser messages correctly.

###Bernoulli Naive Bayes Classifier

In [19]:
#defining the model and training 
model3 = BernoulliNB().fit(X_train, y_train)

#making predictions
y_pred2 = model3.predict(X_test)

#creating a comparison dataframe

comparison_frame = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred2})

comparison_frame.head(5)

Actual  Predicted
0       1          1
1       1          0
2       0          0
3       0          0
4       0          0

In [20]:
#evaluating the model
print(confusion_matrix(y_test, y_pred2))
print(classification_report(y_test, y_pred2))


[[470  45]
 [ 67 260]]
              precision    recall  f1-score   support

           0       0.88      0.91      0.89       515
           1       0.85      0.80      0.82       327

    accuracy                           0.87       842
   macro avg       0.86      0.85      0.86       842
weighted avg       0.87      0.87      0.87       842



this model has the highest accuracy among the three.

##70-30

###Gaussian Naive Bayes Classifier

In [21]:
#dividing data into dependent and independent variable
X1=df.drop(['spam'], axis=1).values
y1=df['spam'].values

#splitting data into train and test set
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.3, random_state=6) 

#defining the model
clf = GaussianNB()  

#training the model
model1 = clf.fit(X1_train, y1_train) 

#making predictions
y1_pred = model1.predict(X1_test)

#creating a comparison dataframe

comparison_frame = pd.DataFrame({'Actual': y1_test, 'Predicted': y1_pred})

comparison_frame.head(5)

Actual  Predicted
0       1          1
1       1          1
2       0          0
3       0          0
4       0          0

In [22]:
#evaluating the model
print(confusion_matrix(y1_test, y1_pred))
print(classification_report(y1_test, y1_pred))

[[547 211]
 [ 23 482]]
              precision    recall  f1-score   support

           0       0.96      0.72      0.82       758
           1       0.70      0.95      0.80       505

    accuracy                           0.81      1263
   macro avg       0.83      0.84      0.81      1263
weighted avg       0.85      0.81      0.82      1263



###Multinomial Naive Bayes Classifier

In [23]:
#defining the model and training 
model2 = MultinomialNB().fit(X1_train, y1_train)

#making predictions
y1_pred1 = model2.predict(X1_test)

#creating a comparison dataframe

comparison_frame = pd.DataFrame({'Actual': y1_test, 'Predicted': y1_pred1})

comparison_frame.head(5)

Actual  Predicted
0       1          0
1       1          0
2       0          0
3       0          0
4       0          0

In [24]:
#evaluating the model

print(confusion_matrix(y1_test, y1_pred1))
print(classification_report(y1_test, y1_pred1))


[[663  95]
 [174 331]]
              precision    recall  f1-score   support

           0       0.79      0.87      0.83       758
           1       0.78      0.66      0.71       505

    accuracy                           0.79      1263
   macro avg       0.78      0.77      0.77      1263
weighted avg       0.79      0.79      0.78      1263



###Bernoulli Naive Bayes Classifier

In [25]:
#defining the model and training 
model3 = BernoulliNB().fit(X1_train, y1_train)

#making predictions
y1_pred2 = model3.predict(X1_test)

#creating a comparison dataframe

comparison_frame = pd.DataFrame({'Actual': y1_test, 'Predicted': y1_pred2})

comparison_frame.head(5)

Actual  Predicted
0       1          1
1       1          0
2       0          0
3       0          0
4       0          0

In [26]:
#evaluating the model
print(confusion_matrix(y1_test, y1_pred2))
print(classification_report(y1_test, y1_pred2))

[[699  59]
 [101 404]]
              precision    recall  f1-score   support

           0       0.87      0.92      0.90       758
           1       0.87      0.80      0.83       505

    accuracy                           0.87      1263
   macro avg       0.87      0.86      0.87      1263
weighted avg       0.87      0.87      0.87      1263



##60-40

###Gaussian Naive Bayes Classifier

In [27]:
#dividing data into dependent and independent variable
X2=df.drop(['spam'], axis=1).values
y2=df['spam'].values

#splitting data into train and test set
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.4, random_state=6) 

#defining the model
clf = GaussianNB()  

#training the model
model1 = clf.fit(X2_train, y2_train) 

#making predictions
y2_pred = model1.predict(X2_test)

#creating a comparison dataframe

comparison_frame = pd.DataFrame({'Actual': y2_test, 'Predicted': y2_pred})

comparison_frame.head(5)

Actual  Predicted
0       1          1
1       1          1
2       0          0
3       0          0
4       0          0

In [28]:
#evaluating the model

print(confusion_matrix(y2_test, y2_pred))
print(classification_report(y2_test, y2_pred))


[[728 275]
 [ 37 644]]
              precision    recall  f1-score   support

           0       0.95      0.73      0.82      1003
           1       0.70      0.95      0.80       681

    accuracy                           0.81      1684
   macro avg       0.83      0.84      0.81      1684
weighted avg       0.85      0.81      0.82      1684



###Multinomial Naive Bayes Classifier

In [29]:
#defining the model and training 
model2 = MultinomialNB().fit(X2_train, y2_train)

#making predictions
y2_pred1 = model2.predict(X2_test)

#creating a comparison dataframe

comparison_frame = pd.DataFrame({'Actual': y2_test, 'Predicted': y2_pred1})

comparison_frame.head(5)

Actual  Predicted
0       1          0
1       1          0
2       0          0
3       0          0
4       0          0

In [30]:
#evaluating the model

print(confusion_matrix(y2_test, y2_pred1))
print(classification_report(y2_test, y2_pred1))

[[856 147]
 [225 456]]
              precision    recall  f1-score   support

           0       0.79      0.85      0.82      1003
           1       0.76      0.67      0.71       681

    accuracy                           0.78      1684
   macro avg       0.77      0.76      0.77      1684
weighted avg       0.78      0.78      0.78      1684



###Bernoulli Naive Bayes Classifier

In [31]:
#defining the model and training 
model3 = BernoulliNB().fit(X2_train, y2_train)

#making predictions
y2_pred2 = model3.predict(X2_test)

#creating a comparison dataframe

comparison_frame = pd.DataFrame({'Actual': y2_test, 'Predicted': y2_pred2})

comparison_frame.head(5)

Actual  Predicted
0       1          1
1       1          0
2       0          0
3       0          0
4       0          0

In [32]:
#evaluating the model
print(confusion_matrix(y2_test, y2_pred2))
print(classification_report(y2_test, y2_pred2))

[[930  73]
 [122 559]]
              precision    recall  f1-score   support

           0       0.88      0.93      0.91      1003
           1       0.88      0.82      0.85       681

    accuracy                           0.88      1684
   macro avg       0.88      0.87      0.88      1684
weighted avg       0.88      0.88      0.88      1684



Bernoulli Naive Bayes Classifier was the best performing model.

changig the test wsize did not make any difference in the accuracy of the models.


#challenging the solution

PCA can be used to select most important features atleast 20 and see if that increases the accuracy

##follow up questions

a). Did we have the right data?

b). Do we need other data to answer our question?

c). Did we have the right question?